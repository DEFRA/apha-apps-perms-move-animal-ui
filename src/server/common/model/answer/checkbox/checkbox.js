import Joi from 'joi'

import { NotImplementedError } from '../../../helpers/not-implemented-error.js'
import { AnswerModel } from '../answer-model.js'
import { validateAnswerAgainstSchema } from '../validation.js'
import { ensureArray } from '../../../helpers/ensure-array.js'

/** @import {AnswerViewModelOptions} from '../answer-model.js' */
/** @import {RawApplicationState} from '~/src/server/common/model/state/state-manager.js' */

/**
 * @param {CheckboxConfig} config
 * @returns {Joi.Schema}
 */
const createCheckboxSchema = (config) => {
  const optionSchema = config.validation.dynamicOptions
    ? Joi.string()
    : Joi.string().valid(...Object.keys(config.options))

  let optionsSchema = Joi.array().required().items(optionSchema)

  if (config.validation.empty) {
    const emptyOptionText = config.validation.empty.message
    optionsSchema = optionsSchema.min(1).messages({
      'array.min': emptyOptionText
    })
  }

  return Joi.object({ [config.payloadKey]: optionsSchema })
}

/**
 * @typedef {{ label: string }} CheckboxOption
 * @typedef {{
 *   payloadKey: string,
 *   options: Record<string, CheckboxOption>,
 *   hint?: string,
 *   isPageHeading? : boolean,
 *   validation: {
 *     empty?: { message: string },
 *     dynamicOptions?: boolean
 *   }
 * }} CheckboxConfig
 */

/**
 * @typedef {string[]} CheckboxData
 */

/**
 * @template Payload
 * @augments AnswerModel<Payload>
 */
export class CheckboxAnswer extends AnswerModel {
  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {CheckboxConfig} */
  get config() {
    return /** @type {any} */ (this.constructor).config
  }

  /** @type {string} */
  get type() {
    const type = 'checkbox'
    return type
  }

  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {CheckboxConfig} */
  static get config() {
    throw new NotImplementedError()
  }

  get value() {
    return this._data !== undefined
      ? ensureArray(this._data?.[this.config.payloadKey])
      : undefined
  }

  get template() {
    return 'model/answer/checkbox/checkbox.njk'
  }

  get html() {
    const data = ensureArray(this._data?.[this.config.payloadKey])

    if (data.length === 0) {
      return 'None'
    }
    return Object.values(data)
      .map((item) => {
        return this.config.options[item].label
      })
      .join('<br /><br />')
  }

  /**
   * @param {AnswerViewModelOptions} options
   */
  async viewModel({ question, validate }) {
    const values = this.value
    const { payloadKey, options, hint } = this.config

    const isPageHeading = this.config.isPageHeading ?? true

    const viewModel = {
      name: payloadKey,
      id: payloadKey,
      fieldset: {
        legend: {
          text: question,
          classes: isPageHeading
            ? 'govuk-fieldset__legend--l'
            : 'govuk-fieldset__legend--m',
          isPageHeading
        }
      },
      items: Object.entries(options).map(([value, option]) => ({
        text: option.label,
        value,
        attributes: {
          'data-testid': `${value}-checkbox`
        },
        checked: (values ?? []).includes(value)
      }))
    }

    if (hint) {
      viewModel.hint = { text: hint }
    }

    if (validate) {
      viewModel.errorMessage = this.validate().errors[this.config.payloadKey]
    }

    return Promise.resolve(viewModel)
  }

  /**
   * @returns {CheckboxData | undefined}
   */
  toState() {
    return ensureArray(this._data?.[this.config.payloadKey])
  }

  validate() {
    const data = this._data?.[this.config.payloadKey]
    return validateAnswerAgainstSchema(createCheckboxSchema(this.config), {
      [this.config.payloadKey]: ensureArray(data)
    })
  }

  /**
   * @param {Payload} fields
   */
  _extractFields(fields) {
    return /** @type {Payload} */ ({
      [this.config.payloadKey]: fields[this.config.payloadKey]
    })
  }

  /**
   * @param {string[] | undefined} state
   * @param {RawApplicationState} [context]
   * @returns {CheckboxAnswer}
   */
  static fromState(state, context) {
    return new this(
      state !== undefined ? { [this.config.payloadKey]: state } : undefined,
      context
    )
  }
}
