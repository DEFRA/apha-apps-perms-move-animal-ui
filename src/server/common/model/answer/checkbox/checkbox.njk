{% from "govuk/components/checkboxes/macro.njk" import govukCheckboxes %}

{{ govukCheckboxes(answer.viewModel(viewModelOptions)) }}
