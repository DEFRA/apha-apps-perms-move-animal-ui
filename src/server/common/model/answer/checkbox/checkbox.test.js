import { CheckboxAnswer } from './checkbox.js'

/** @import {CheckboxConfig} from './checkbox.js' */

const checkboxEmptyError = 'Select at least one checkbox'

/** @type {CheckboxConfig} */
const config = {
  payloadKey: 'test_checkbox',
  hint: 'test_hint',
  options: {
    badgerProofFencing: {
      label:
        'Badger proof fencing, such as solid aluminium sheeted gates, aluminium sheeting on rail fences, retractable electric fences'
    },
    limitAccessToBadgerHabitat: {
      label: 'Limiting access to badger latrines and setts'
    }
  },
  errors: {
    emptyOptionText: { message: checkboxEmptyError }
  }
}

class TestCheckboxAnswer extends CheckboxAnswer {
  static config = config
}

class TestCheckboxAnswerAllOptional extends CheckboxAnswer {
  static config = { ...config, errors: {} }
}

const validTestCheckbox = { test_checkbox: ['badgerProofFencing'] }

describe('Checkbox.new', () => {
  it('should strip away any irrelevant values', () => {
    const payload = { ...validTestCheckbox, nextPage: '/other/page' }
    const testInstance = new TestCheckboxAnswer(payload)

    expect(testInstance._data).toEqual(validTestCheckbox)
  })
})

describe('#Checkbox.validate', () => {
  it('should return true for valid value (passed in as a single answer)', () => {
    const { isValid, errors } = new TestCheckboxAnswer({
      test_checkbox: 'badgerProofFencing'
    }).validate()

    expect(isValid).toBe(true)
    expect(errors).toEqual({})
  })

  it('should return true for valid value (passed in as an array)', () => {
    const { isValid, errors } = new TestCheckboxAnswer({
      test_checkbox: ['badgerProofFencing']
    }).validate()

    expect(isValid).toBe(true)
    expect(errors).toEqual({})
  })

  it('should return true for another valid value', () => {
    const { isValid, errors } = new TestCheckboxAnswer({
      test_checkbox: 'limitAccessToBadgerHabitat'
    }).validate()

    expect(isValid).toBe(true)
    expect(errors).toEqual({})
  })

  it('should return true for an array of valid values', () => {
    const { isValid, errors } = new TestCheckboxAnswer({
      test_checkbox: ['badgerProofFencing', 'limitAccessToBadgerHabitat']
    }).validate()

    expect(isValid).toBe(true)
    expect(errors).toEqual({})
  })

  it('should return false for empty', () => {
    const { isValid, errors } = new TestCheckboxAnswer({
      test_checkbox: []
    }).validate()

    expect(isValid).toBe(false)
    expect(errors.test_checkbox.text).toBe(checkboxEmptyError)
  })

  it('should return false for invalid values, even if mixed with valid values', () => {
    const { isValid } = new TestCheckboxAnswer({
      test_checkbox: ['notAValue', 'badgerProofFencing']
    }).validate()

    expect(isValid).toBe(false)
    // no mandated error message here, because this should only be hit if user
    // manually constructs a POST payload
  })

  describe('TestCheckboxAnswerAllOptional', () => {
    it('should not error if no options are selected', () => {
      const { isValid } = new TestCheckboxAnswerAllOptional({
        test_checkbox: []
      }).validate()

      expect(isValid).toBe(true)
    })

    it('should not error if the question has not yet been answered', () => {
      const { isValid } = new TestCheckboxAnswerAllOptional(
        undefined
      ).validate()

      expect(isValid).toBe(true)
    })
  })
})

describe('#Checkbox.toState', () => {
  it('should return an object with the correct data', () => {
    const checkbox = new TestCheckboxAnswer({
      test_checkbox: ['badgerProofFencing']
    })
    expect(checkbox.toState()).toEqual(['badgerProofFencing'])
  })

  it('should normalise a single option into an array', () => {
    const checkbox = new TestCheckboxAnswer({
      test_checkbox: 'badgerProofFencing'
    })
    expect(checkbox.toState()).toEqual(['badgerProofFencing'])
  })

  it('should return an object with the correct data (with more than one selected)', () => {
    const confirmation = new TestCheckboxAnswer({
      test_checkbox: ['badgerProofFencing', 'limitAccessToBadgerHabitat']
    })
    expect(confirmation.toState()).toEqual([
      'badgerProofFencing',
      'limitAccessToBadgerHabitat'
    ])
  })
})

describe('Checkbox.fromState', () => {
  it('should re-construct the payload from a valid state', () => {
    const state = new TestCheckboxAnswer(validTestCheckbox).toState()
    expect(TestCheckboxAnswer.fromState(state)._data).toEqual(validTestCheckbox)
  })

  it('should return an undefined value if the state is undefined', () => {
    expect(TestCheckboxAnswer.fromState(undefined)._data).toBeUndefined()
  })
})

describe('Checkbox.value', () => {
  it('should return a value-wrapped object to rendering in the template', () => {
    expect(
      new TestCheckboxAnswer({ test_checkbox: ['badgerProofFencing'] }).value
    ).toEqual(['badgerProofFencing'])
  })

  it('should return undefined if the payload is undefined', () => {
    expect(new TestCheckboxAnswer(undefined).value).toBeUndefined()
  })

  it('should return an array-wrapped payload, even if the initial value is a string', () => {
    expect(
      new TestCheckboxAnswer({ test_checkbox: 'badgerProofFencing' }).value
    ).toEqual(['badgerProofFencing'])
  })
})

describe('CheckboxAnswer.template', () => {
  it('should return the radio button model template', () => {
    const radio = new TestCheckboxAnswer(validTestCheckbox)
    expect(radio.template).toBe('model/answer/checkbox/checkbox.njk')
  })
})

describe('CheckboxAnswer.html', () => {
  it('should return the payload with HTML', () => {
    const checkbox = new TestCheckboxAnswer({
      test_checkbox: ['badgerProofFencing', 'limitAccessToBadgerHabitat']
    })

    expect(checkbox.html).toMatchSnapshot()
  })

  it('should return None for undefined state', () => {
    const checkbox = new TestCheckboxAnswer(undefined)
    expect(checkbox.html).toBe('None')
  })

  it('should return None for an empty list', () => {
    const checkbox = new TestCheckboxAnswer({
      test_checkbox: []
    })
    expect(checkbox.html).toBe('None')
  })
})

describe('CheckboxAnswer.viewModel', () => {
  const question =
    'Which measures are you taking to reduce contamination from wildlife?'

  const defaultViewModel = {
    name: 'test_checkbox',
    id: 'test_checkbox',
    hint: {
      text: 'test_hint'
    },
    fieldset: {
      legend: {
        text: question,
        isPageHeading: true,
        classes: 'govuk-fieldset__legend--l'
      }
    }
  }

  const itemOne = {
    value: 'badgerProofFencing',
    text: TestCheckboxAnswer.config.options.badgerProofFencing.label,
    attributes: {
      'data-testid': 'badgerProofFencing-checkbox'
    },
    checked: false
  }

  const itemTwo = {
    value: 'limitAccessToBadgerHabitat',
    text: TestCheckboxAnswer.config.options.limitAccessToBadgerHabitat.label,
    attributes: {
      'data-testid': 'limitAccessToBadgerHabitat-checkbox'
    },
    checked: false
  }

  it('should return everything (except errors) to render in the template', () => {
    const validAnswer = new TestCheckboxAnswer({
      test_checkbox: ['limitAccessToBadgerHabitat']
    })

    const expectedItems = [itemOne, { ...itemTwo, checked: true }]

    expect(validAnswer.viewModel({ validate: false, question })).toEqual({
      ...defaultViewModel,
      items: expectedItems
    })
  })

  it('should return everything (including errors) to render in the template', () => {
    const invalidAnswer = new TestCheckboxAnswer({
      test_checkbox: []
    })

    const expectedItems = [itemOne, itemTwo]

    expect(invalidAnswer.viewModel({ validate: true, question })).toEqual({
      ...defaultViewModel,
      errorMessage: { text: checkboxEmptyError },
      items: expectedItems
    })
  })

  it('should return data to render with the alternative (not page heading) question styles set correctly', () => {
    /** @type {CheckboxConfig} */
    const checkboxConfigNotPageHeading = {
      ...config,
      isPageHeading: false
    }

    class NonPageHeadingCheckboxAnswer extends CheckboxAnswer {
      static config = checkboxConfigNotPageHeading
    }
    const nonPageHeadingAnswer = new NonPageHeadingCheckboxAnswer({
      test_checkbox: ['limitAccessToBadgerHabitat']
    })

    const expectedItems = [itemOne, { ...itemTwo, checked: true }]

    expect(
      nonPageHeadingAnswer.viewModel({ validate: false, question })
    ).toEqual({
      ...defaultViewModel,

      fieldset: {
        legend: {
          ...defaultViewModel.fieldset.legend,

          classes: '',
          isPageHeading: false
        }
      },
      items: expectedItems
    })
  })
})
