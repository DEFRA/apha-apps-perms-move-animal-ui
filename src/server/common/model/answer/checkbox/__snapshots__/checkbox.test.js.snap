// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxAnswer.html should return the payload with HTML 1`] = `"Badger proof fencing, such as solid aluminium sheeted gates, aluminium sheeting on rail fences, retractable electric fences<br /><br />Limiting access to badger latrines and setts"`;
